### DISCUSSION TITLE
/hyndsight/arimaconstants/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#151] [Announcements]
Constants and ARIMA models in R 
 https://robjhyndman.com/hyndsight/arimaconstants/

**Comment by robjhyndman** [comment]
```header
Originally posted by syazreen on 2013-11-11 03:16:58
````
Hi Prof,\
auto.arima function is great, the best model can automatically be fitted
.by having smallest error.\
Hpwever my model required a random walk with drift model.\
How to use arima() to include drift term.\
arima(x,c(0,1,0)) will not give the same model as I want it to be; that
is random walk with drift term is the mean.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-11-11 03:29:39
````
auto.arima() returns the fitted model.

If you need to fit it again, use\
Arima(x, order=c(0,1,0), include.drift=TRUE)

**Reply by robjhyndman** [reply]
```header
Originally posted by syazreen on 2013-12-02 20:47:25
````
Thanks so much!\
It works exactly as what I want.

**Comment by robjhyndman** [comment]
```header
Originally posted by Brian on 2013-12-07 04:28:46
````
Is it possible the behavior of Arima has changed?

install.packages(&quot;fma&quot;)\
library(fma)

mod\&lt;-Arima(eggs,c(1,3,0),include.drift=TRUE)\
mod

still includes a drift term even though d \&gt;2.

Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-12-07 10:01:10
````
Well spotted. The drift term is actually redundant and non-identifiable
\-- note the size of the standard error on the drift coefficient. So it
is not actually fitting a cubic drift. I\'ll fix the function so it
doesn\'t return the parameter.

**Comment by robjhyndman** [comment]
```header
Originally posted by sunsetter on 2014-01-28 05:53:29
````
Can auto.arima() be set to never allow non-zero mean, even when d=0?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-01-28 09:13:08
````
No. You would have to do your own modification of it if you wanted to do
that.

**Reply by robjhyndman** [reply]
```header
Originally posted by Erdal on 2016-05-24 18:56:45
````
hello Mr. Hyndman,\
can auto.arima be set to never allow zero mean for case d=0?

**Comment by robjhyndman** [comment]
```header
Originally posted by Débora Spenassato on 2014-06-24 18:59:45
````
Hi prof. Hyndman,\
my model using the function auto. arima() is ARIMA(1,1,0) with drift (AR
= 0.208 and drift =2.531). I\'m having difficulty to form the equation.
would be deltaYt=2.531+0.208 deltaYt-1 + et?

Thanks,

**Comment by robjhyndman** [comment]
```header
Originally posted by Shraddha Panda on 2014-07-30 16:59:11
````
Hello Professor, Could you please have a look at this question here and
share your inputs? I am trying to fit auto.arima for longitudnal data by
grouping different regions..
[http://stackoverflow.com/qu\...](http://stackoverflow.com/questions/25036986/auto-arima-using-xreg-and-forecasting-several-ts-together &quot;http://stackoverflow.com/questions/25036986/auto-arima-using-xreg-and-forecasting-several-ts-together&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by Joshua Makubu on 2015-03-29 16:57:34
````
Hi Prof\
Arima() is not a function in R is the feedback i get when i try to model
with drift. Please educate me

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-03-29 22:50:21
````
load the forecast package first.

**Reply by robjhyndman** [reply]
```header
Originally posted by Joshua Makubu on 2015-04-02 22:46:12
````
Prof .Good evening, Am using r version 3.1.2 but still cant get the
function auto.arima or the function arima after installing the forecast
package.\
is there anything am not doing right?

**Reply by robjhyndman** [reply]
```header
Originally posted by Mathijs on 2015-04-08 08:04:43
````
It\'s likely you haven\'t loaded the package, load the package by
entering library(forecast). After that, loading the help files by
entering ?auto.arima (use R Studio!) and the examples in Prof.
Hyndman\'s book will help you get there.

**Comment by robjhyndman** [comment]
```header
Originally posted by JJJJJ on 2015-12-07 10:42:49
````
Does the same apply when i include dummy variables? My model seems to
get a trend when i include dummy variables (together with my
differencing).

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-12-07 10:45:43
````
The dummy variables should be differenced before use. Otherwise they
will induce a trend.

**Reply by robjhyndman** [reply]
```header
Originally posted by Alex Mustermann on 2015-12-07 10:51:42
````
thanks!

**Comment by robjhyndman** [comment]
```header
Originally posted by Ashwini Srinivas on 2016-05-04 05:56:07
````
Hi Rob,\
I have used auto.arima and the best model is 1,1,1 and there is no
mention of zero or non zero mean and also my forecast fallows the
straight line could you please throw some light on this

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-05-04 06:11:28
````
It sounds like you have a drift term. In this case, the slope is the
constant.

**Reply by robjhyndman** [reply]
```header
Originally posted by Ashwini Srinivas on 2016-05-04 07:09:12
````
Thanks a lot .I am really sorry to ask this what exactly is the drift
term i am unable to find much on net regarding this

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-05-04 22:49:21
````
Read the post.

**Comment by robjhyndman** [comment]
```header
Originally posted by Erdal on 2016-05-24 19:28:08
````
hello Mr. Hyndman,

can auto.arima be set to never allow zero mean for case d=0?\
ı want it to give only models with intercept. Adding allowmean=FALSE
give only zero mean models but allowmean=TRUE gives both zero and
nonzero mean models. &quot;allowmean: If TRUE, models with a non-zero mean
are considered.&quot; this expresion does not work.\
how can i run function with only nonzero mean models or fix intercept
for all models run by function?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-05-24 21:04:01
````
No. But you can specify your own model using Arima() with
include.mean=TRUE

**Reply by robjhyndman** [reply]
```header
Originally posted by Erdal on 2016-05-31 13:08:54
````
Thank you Mr. Hyndman.\
It resulted in models with intercept. But all models are ARIMA(0,0,0).
Is it possible?\
Data is in form of first difference.\
code is:\
(Arima(ts(y, freq=143), include.mean=TRUE,
xreg=as.data.frame.list(v19\[i,\]\[ !v19\[i,\] %in% Remove\]))

**Reply by robjhyndman** [reply]
```header
Originally posted by Erdal on 2016-06-02 09:47:37
````
Hi again Mr. Hyndman.

I want to get the models with intercept and automatic ARIMA(p,d,q)
values.\
This is only possible with auto.arima function.\
But allowmean=T does not work.\
Is it a problem or a bug?\
It is said in Package 'forecast' (April 14, 2016) that when allowmean=T
models with a non-zero mean are considered.

Can it be fixed?

This is very important for me as i am writting my master\'s thesis. I
can send you the deatil of the code i use in study.\
Thank you very much.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-06-02 10:04:04
````
There is no bug. allowmean allows a mean as the help file explains. It
doesn\'t force a non-zero mean. If you want to force a specific model,
use Arima(). Don\'t send me code \-- I do not provide a help service to
the whole world.

**Reply by robjhyndman** [reply]
```header
Originally posted by Erdal on 2016-06-02 10:56:41
````
Thank you Mr. Hyndman.\
I thought that it will also be helpful for the other researchers to run
the auto.arima() with only nonzero mean models/force a non-zero mean.

Allowmean is a recent addition to the function (Version 6.0 (9 May
2015)).\
It works very well for only zero mean models. It also gives results with
both zero and non zero models.

It can be helpful to set it for only non-zero mean models (force
non-zero mean).

I do not know whether this is possible or not/logical or not?

Thanks a lot.

(I also write it in
[http://stackoverflow.com/qu\...](http://stackoverflow.com/questions/37422790/how-to-run-auto-arima-for-only-nonzero-mean-models &quot;http://stackoverflow.com/questions/37422790/how-to-run-auto-arima-for-only-nonzero-mean-models&quot;)
)\
But i could\'t get an answer.)

**Comment by robjhyndman** [comment]
```header
Originally posted by Anne on 2016-08-26 09:36:27
````
Hi Mr. Hyndman,\
I defined a model using your nice auto.arima function and I include
dummy variables for day of the week in xreg. When I use only 6 out of 7
week-dummy variables (to have a full rank matrix) the model works well.
However, I would like to include all 7 days of the week as dummy
variables. Hence, I need to exclude the constant/intercept in the model
if I understand correctly. When I use Arima with include.constant=F and
using several different orders of p, d and q I get the error: Error in
optim(init\[mask\], armaCSS, method = optim.method, hessian = FALSE,
:non-finite value supplied by optim. What could be causing this error? I
would like to understand why the model does work with only 6 out of 7
week-dummies and does not work when I include all 7 week-dummies and
exclude the intercept.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-08-27 07:57:50
````
You would to provide a minimal reproducible example. The following code
works.\

    z &lt;- ts(rnorm(200), freq=7)
    xreg &lt;- seasonaldummy(z)
    xreg &lt;- cbind(xreg, 1-rowSums(xreg))
    Arima(z, xreg=xreg, order=c(1,0,1), include.constant=FALSE)

**Comment by robjhyndman** [comment]
```header
Originally posted by Shanu Agrawal on 2016-11-11 11:42:38
````
I am running ARIMA model with external regressors, but getting error:
&quot;NO Suitable ARIMA model found&quot;.\
What might be the reasons?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-11-11 23:21:32
````
The regressors could be rank deficient.

**Comment by robjhyndman** [comment]
```header
Originally posted by Shinta Ps on 2017-02-15 06:26:13
````
hi, Prof\
how to predict arima with outlier and constant ?\
can you help me please\
thank you :)

**Comment by robjhyndman** [comment]
```header
Originally posted by Seyi Ajao on 2017-03-16 03:15:17
````
Dear Prof. Please how do I write my model ARIMA(0,1,0) with drift 7.0409
and s.e. 3.5851?

**Comment by robjhyndman** [comment]
```header
Originally posted by Andrzej on 2017-03-23 10:41:58
````
Hi Profesor,\
it might be a stupid question but how does the Backshift operator
interacts with variable t (time). If I want to derive the forecasting
equation lets say from the last equation you present within this post
(showing case of d = 1 and include.drift = TRUE), i need to multiple
expressions in the first two parentheses and I am not sure whether the
backshift operator applied to variable t (time) should result in
\'lagged time\': B\*t = t-1, or no: B\*t = t.

**Comment by robjhyndman** [comment]
```header
Originally posted by prakash chandra on 2017-06-12 12:36:06
````
Hello sir, I have a problem in understanding p &amp; q for ARIMA. and how we
calculate what should be value of the lag.max in ACF &amp; PACF .\
Please sir give some suggestion in this matter. I shall be obliged to
you for this.

**Comment by robjhyndman** [comment]
```header
Originally posted by Duaa Taleb on 2017-10-22 18:28:00
````
Hi Prof,\
when i use arima() i have AIC differance from Arima() for the same
data.\
\-\-\-\--\
\&gt;Arima(x, order=c(0,1,1), include.drift=TRUE)\
Series: x\
ARIMA(0,1,1) with drift\
Coefficients:\
ma1 drift\
-0.2021 3.4921\
s.e. 0.0960 5.3835\
sigma\^2 estimated as 4757: log likelihood=-581.24\
AIC=1168.48 AICc=1168.72 BIC=1176.41\
\-\-\-\-\-\-\-\-\--\
\&gt; arima(x,c(0,1,1))\
Call:\
arima(x = x, order = c(0, 1, 1))\
Coefficients:\
ma1\
-0.1971\
s.e. 0.0952\
sigma\^2 estimated as 4684: log likelihood = -587.09, aic = 1178.19\
\-\-\-\-\-\-\-\-\-\-\--\
Which function is the best? and what the differance between them?\
Thanks

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-10-22 20:59:26
````
You are fitting two different models, one with drift and one without.
The aic is the same from both functions if you compare the same model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Fernando Adventius on 2017-11-16 16:20:03
````
Hi Prof Hyndman, I tried to understand your Statistical Package in R.\
Let say, we have an AR(1) model i.e.,
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/2822d485f2d44069a55a766ddfb0d3619afb3e1850d6698517443e1805885090.gif)\
And, the ARIMA in R function gives me the same results.\
My question is: for t = 0 that is the very first sample in the Time
Series, how do you estimate the ARIMA sample estimate for that point?
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/86eacb53fe0c076db96bd18dcfcd407b3dbb9e4f6a6561d021713683a1c97586.gif)\
How do you determine the sample estimate for t = 0?\
This is particularly important to determine the sigma\^2 and also the
AICC of the ARIMA Model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Vivek N K on 2017-12-04 07:29:26
````
How to include a negative drift/trend? When we remove seasonality and
trends, but add allow.drift=TRUE, i get uptrend in some time-series.
But, for time-series with negative trend (i.e. down trend), it is
ignored. i.e. I do not get a down trending forecast in all cases. I even
tried giving negative values for D (-1 to -3) in auto.arima function. is
there a separate approach for implementing down-trending time-series?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-12-04 07:36:41
````
Trends are estimated from the data, not imposed on them.

**Reply by robjhyndman** [reply]
```header
Originally posted by Vivek N K on 2017-12-04 08:04:56
````
I agree. I could not use arima of order (0,1,1) since, I have regressors
(xreg - 5 variables), which will identify the correct trends in
forecast. So, I had to have both regressors and drift trends (+ve and
-ve). Is this possible in Arima?

**Comment by robjhyndman** [comment]
```header
Originally posted by Nibir Nath on 2017-12-05 06:20:09
````
Hi Prof,\
How to calculate the ideal d value for a seasonal ARIMA model?

**Comment by robjhyndman** [comment]
```header
Originally posted by Tanmay on 2018-02-05 07:20:07
````
Hello Sir,\
I used auto.arima function to forecast the values. Do I have to
difference and transform the data first before running auto.arima.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-02-05 09:01:03
````
No. Read
[https://www.otexts.org/fpp2\...](https://www.otexts.org/fpp2/arima-r.html &quot;https://www.otexts.org/fpp2/arima-r.html&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by olivier vermeersch on 2018-04-03 13:57:48
````
Hi Professor,\
I\'ve fitted a model with the auto.arima function (default options) and
a model with an intercept and and 1 xreg variable have been proposed.

Series: training\
Regression with ARIMA(0,0,0) errors

Coefficients:\
intercept acquisition_lag11\
463.0352 0.0272\
s.e. 68.0278 0.0094

sigma\^2 estimated as 14979: log likelihood=-372.55\
AIC=751.1 AICc=751.53 BIC=757.39

However, when I want to apply the model (see below) I receive an error:

apply_model\&lt;-Arima(target_ts,xreg=predictors_ts\[,modelbuilding_regressors\],model=modelbuilding_fit)

Error in if (ncol(xreg) != ncol(model\$xreg)) stop(&quot;Number of
regressors does not match fitted model&quot;) :\
argument is of length zero

If I run predictors_ts\[,modelbuilding_regressors\] I get, which seems
ok to me (1 xreg).

Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec\
2012 4394 4343 4882 5448 6084 5323 5290 6496 5624 5933 5748 5081\
2013 5441 5187 5505 5272 6756 9097 9648 10316 9431 9956 12067 9659\
2014 9854 7739 8081 9388 7878 7896 8244 7938 6932 7640 8379 6694\
2015 7273 6847 6834 7017 6221 7563 7776 7465 6322 7215 10034 4198\
2016 6373 6335 5981 7153 6356 5810 6994 5959 6302 6383 7251 7524\
2017 6486 6337 6748 6335 6049 6295 6358 5793 5954 6462 10470 5410

Any idea what is happening here as I only have 1xreg so it should work?\
Thx for any feedback.

Olivier

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-04-03 22:36:58
````
This is not a help site. Ask on
[stackoverflow.com](http://stackoverflow.com &quot;stackoverflow.com&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by Karen on 2018-11-04 11:31:20
````
Hi, Professor Hyndman, does the auto.arima function remove outliers?
Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-11-04 20:58:37
````
No. Use tsclean to remove outliers before running auto.arima. See
[https://otexts.org/fpp2/mis\...](https://otexts.org/fpp2/missing-outliers.html#outliers &quot;https://otexts.org/fpp2/missing-outliers.html#outliers&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted by Karen on 2018-11-04 21:04:11
````
Thanks Professor Hyndman!

**Comment by robjhyndman** [comment]
```header
Originally posted by Airlane Pereira Alencar on 2019-04-23 13:29:08
````
Thanks for your blog and your book.\
I am fitting a SARIMA (1,0,0)(0,1,2)\_4 model with drift for quarterly
production data and I am not sure if the model is
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/b375f0d3f1b6a18041662462ddbba5a90aa4a91032f5e28cd77bc541ab99f170.gif)\
I checked calculating the forecasts with this model and the results were
close to the obtained with the forecast for the Arima fit.\
Is this the corresponding model?\
Thanks\
Airlane\
[www.ime.usp.br/\~lane](http://www.ime.usp.br/~lane &quot;www.ime.usp.br/~lane&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-04-23 21:48:45
````
I don\'t have the capacity to run a help service. Please ask on
[crossvalidated.com](http://crossvalidated.com &quot;crossvalidated.com&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by Dennis on 2019-05-01 22:22:43
````
Hello prof. Hyndman,

I have two questions about the auto.arima() function.

1\. Is there a possibility to get the value of the constant c for a
model fitted by auto.arima() ?

I would like to get a better understanding on how the point forecasts
are produced but sometimes get different values. whenever I try to
follow the formula c=μ(1−ϕ1−⋯−ϕp) with μ = (1−B)\^d \*yt I get weird
results.

2\. When a drift is used in an auto.arima() model, how does the drift
value effect the formula of the point forecasts?

Thank you very much

BR

Dennis

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-05-02 02:43:17
````
When there is no differencing, c=μ(1−ϕ1−⋯−ϕp) and μ = an estimate of the
mean of the training data. When there is a drift term (with d=1 and
D=0), the drift parameter is μ and estimate of the mean of the first
differenced data. Note that the estimate of the mean in each case is not
necessarily the sample mean.

The drift term induces a linear trend in the forecasts.

See
[https://otexts.com/fpp2/ari\...](https://otexts.com/fpp2/arima-r.html#understanding-constants-in-r &quot;https://otexts.com/fpp2/arima-r.html#understanding-constants-in-r&quot;)
for more details.

**Reply by robjhyndman** [reply]
```header
Originally posted by Dennis on 2019-05-03 11:42:56
````
thanks for quick response.

But how is μ estimated in the case d=0 / D=1? Is μ then the mean of the
seasonal differenced data?\
And what about the case d=1/D=1?

You said that the estimate of the mean in each case is not necessarily
the sample mean. Do I understand correctly? μ is optimised by Maximum
likelihood and therefore it can have a difference to the sample mean?

Thank you very much\
I really appreciate your time

BR\
Dennis

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-05-03 23:54:08
````
Yes, when d=0, D=1, it is an estimate of the mean of the seasonally
differenced data. When d=D=1, it is an estimate of the mean of the
seasonally and first differenced data.

Yes, MLE is used.

**Comment by juancarlosfloresr** [comment]
Hi Mr. Hyndman

If i run auto.arima funtion to a seasonal time series, can i get a SARIMA model with drift? Eg. (0,0,1)(1,1,0)[12] with drift.

The reason of my question is because the drift term is usually associated with non-stationary series.

**Reply by juancarlosfloresr** [reply]
Correction: ...associated with non-seasonal data.

**Reply by robjhyndman** [reply]
Yes, you can have a drift term with a seasonal ARIMA model.

**Reply by juancarlosfloresr** [reply]
Thanks a lot.

In which chapter of your book can I find theory about it?

Note: Im reading "Forecasting: Principles and Practice (3rd ed)"

Thanks!

**Reply by robjhyndman** [reply]
https://otexts.com/fpp3/seasonal-arima.html Just add a constant to the equation to get a drift term.

**Comment by mariospelekanos** [comment]
Dear Professor,

if a model chosen by auto.arima fails Ljung-Box test (p-value<0.05 using checkresiduals) and this happens also for other best AIC alternatives for SARIMA and SARIMAX forecasting models then what is the best we can do? tsoutliers and Box-Cox transformation are already used. Even some observations in the beginning have been omitted for reducing volatility. It is noted that long-term forecasting is needed. 
Thank you in advance.

**Reply by robjhyndman** [reply]
Either find another model class, or live with an imperfect model.

**Comment by AndGatos** [comment]
Hello Professor Hyndman, 

I have a question regarding the functions "Arima" and "auto.arima". 

I have read in your textbook that:
> The Arima() function will never return a model with inverse roots outside the unit circle. The auto.arima() function is even stricter, and will not select a model with roots close to the unit circle either. 

I have trained 2 models with the same input data, one with Arima (let it be "model") and one with auto.arima (let it be "auto_model"). 
The p-value and AICc are shown in the following table. 

 | Models       | p-value   |  AICc   |
| -------------- | ----------- | ---------|
| model          | 0.52        | 949.64 |
| auto_model | 0.019      | 998.41 | 


After that, I plotted the inverse characteristic roots for each model using the "autoplot" function. 

<a href="https://ibb.co/KNr04Zv"><img src="https://i.ibb.co/VHqjny4/the-model.png" alt="the-model" border="0"></a>

<a href="https://ibb.co/tBTS4HN"><img src="https://i.ibb.co/Q93ZfJq/the-auto-model.png" alt="the-auto-model" border="0"></a>

If I had to choose between these models, which one would be preferable? 
Can it be checked by cross validation, if a model becomes unstable?

Thank you very much for your assistance.

**Reply by robjhyndman** [reply]
The first one has roots close to the unit circle. So it will be numerically unstable. Use the second one. Yes, you could check the forecasts using time series CV.

**Reply by AndGatos** [reply]
Thank you very much Professor!
